include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -0,0 +1,1265 @@
+            if (operational_status == operational_domain::operational_status::OPERATIONAL)
+            {
+                queue_next_points(sp);
+            }
+        }
+
+        log_stats();
+
+        return op_domain;
+    }
+    /**
+     * Performs contour tracing to determine the operational domain. The algorithm first performs a random sampling of
+     * up to the specified number of samples. It stops random sampling once it finds a single operational point, from
+     * which it moves straight outwards until it encounters the counter of the operational domain. From this point, it
+     * traces the contour until it reaches the initial contour point again. The operational domain will finally only
+     * contain up to `samples` random non-operational points as well as the contour of the found operational domain plus
+     * a one pixel wide border around it.
+     *
+     * @param samples Maximum number of random samples to be taken before contour tracing.
+     * @return The (partial) operational domain of the layout.
+     */
+    [[nodiscard]] operational_domain contour_tracing(const std::size_t samples) noexcept
+    {
+        mockturtle::stopwatch stop{stats.time_total};
+
+        // first, perform random sampling to find an operational starting point
+        const auto starting_point = find_operational_step_point_via_random_sampling(samples);
+
+        // if no operational point was found within the specified number of samples, return
+        if (!starting_point.has_value())
+        {
+            return op_domain;
+        }
+
+        const auto next_clockwise_point = [](std::vector<step_point>& neighborhood,
+                                             const step_point&        backtrack) noexcept -> step_point
+        {
+            assert(std::find(neighborhood.cbegin(), neighborhood.cend(), backtrack) != neighborhood.cend() &&
+                   "The backtrack point must be part of the neighborhood");
+
+            while (neighborhood.back() != backtrack)
+            {
+                std::rotate(neighborhood.begin(), neighborhood.begin() + 1, neighborhood.end());
+            }
+
+            return neighborhood.front();
+        };
+
+        // find an operational point on the contour starting from the randomly determined starting point
+        const auto contour_starting_point = find_operational_contour_step_point(*starting_point);
+
+        auto current_contour_point = contour_starting_point;
+        auto backtrack_point       = current_contour_point.x == 0 ?
+                                         current_contour_point :
+                                         step_point{current_contour_point.x - 1, current_contour_point.y};
+
+        auto current_neighborhood = moore_neighborhood(current_contour_point);
+        auto next_point           = next_clockwise_point(current_neighborhood, backtrack_point);
+
+        while (next_point != contour_starting_point)
```

Comment:
As you mentioned, contour tracing is not yet applicable to gates where we have disconnected domains. Can it happen that the algorithm never stops? Especially since we can go from one domain to another with ``find_operational_contour_step_point``?